File: LeetCode/Problems/Python/113. Path Sum II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#Iterative DFS
#Time: O(n)
#Space:O (n)
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root:
            return root
        
        stack = [(root, sum, [root.val])]
        result = []
        
        while stack:
            node, remain, path = stack.pop()
            remain -= node.val
            
            #if we found a leaf
            if not node.left and not node.right:
                if remain == 0:
                    result.append(path)
            if node.left:
                stack.append((node.left, remain, path + [node.left.val]))
            if node.right:
                stack.append((node.right, remain, path + [node.right.val]))
            
        return result
#dfs + backtracking 
#Time: O(n)
#Space:O (n)
class Solution:
    def dfs(self, node, remain, path):
        if not node:
            return
        #print(node.val, remain)
        remain -= node.val
        path.append(node.val)
        if not node.left and not node.right and remain == 0:
            #copy the found path
            self.result.append(path[:])
        if node.left:
            self.dfs(node.left, remain, path)
        if node.right:
            self.dfs(node.right, remain, path)
        #backtracking
        path.pop()
        return
        
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root:
            return root

File: LeetCode/Problems/C++/113. Path Sum II.cpp
//post order iterative
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> paths;
        vector<int> found_path;
        stack<TreeNode*> st;
        TreeNode* node = root;
        TreeNode* prev = NULL;
        while(node ||!st.empty()){
            if(node){
                st.push(node);
                found_path.push_back(node->val);
                sum -= node->val;
                node = node->left;  
            }
            else{
                node = st.top();
                if(node->right && prev != node->right){
                    //traverse right subtree
                    node = node->right;
                }
                else{
                    //leaf
                    if(!node->left && !node->right && sum ==0)
                        paths.emplace_back(found_path);
                    //backtracking
                    sum += node->val;
                    found_path.pop_back();
                    st.pop();
                    prev = node;
                    node = NULL;
                }
            }  
        }
        return paths;
    }
};


class Solution {
public:
    void findPath(TreeNode* node, vector<vector<int>>& paths, vector<int> path, int sum){
        if(node==NULL){
            return;
        }
        sum-= node->val;
        path.push_back(node->val);
        if(node->left == NULL && node->right == NULL && sum == 0){
            paths.push_back(path);
        }
        
        findPath(node->left, paths, path, sum);
        findPath(node->right, paths, path, sum);
    }
    
    
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> paths;
        
        findPath(root, paths, {}, sum);
        
        return paths;
    }
};